src/saveload/afterload.cpp

```diff
@@ -3241,6 +3241,50 @@ bool AfterLoadGame()
 		}
 	}
 
+	if (IsSavegameVersionBefore(SLV_CARGO_MOVEMENT)) {
+		/* Find a reasonable source and travel distance for cargo in vehicles. */
+		for (Vehicle *v : Vehicle::Iterate()) {
+			for (auto it = v->cargo.Packets()->begin(); it != v->cargo.Packets()->end(); it++) {
+				auto first_station_id = (*it)->GetFirstStation();
+				auto last_station_id = v->First()->last_loading_station;
+
+				if (first_station_id == INVALID_STATION && last_station_id == INVALID_STATION) {
+					/* No station to calculate distance traveled. Just start counting from the current tile. */
+					(*it)->UpdateLoadingTile(v->tile);
+				} else if (first_station_id == INVALID_STATION) {
+					auto last_station = Station::Get(last_station_id);
+
+					/* We don't know where this cargo started, so let's just say it started from the last station. */
+					(*it)->UpdateLoadingTile(last_station->xy);
+				} else if (last_station_id == INVALID_STATION) {
+					auto first_station = Station::Get(first_station_id);
+
+					/* We don't know where the train was last, so just assume we were picking it up at the source. */
+					(*it)->UpdateLoadingTile(first_station->xy);
+				} else {
+					auto first_station = Station::Get(first_station_id);
+					auto last_station = Station::Get(last_station_id);
+
+					/* Act like we loaded it at the first station and unloaded at the last. */
+					(*it)->UpdateLoadingTile(first_station->xy);
+					(*it)->UpdateUnloadingTile(last_station->xy);
+					(*it)->UpdateLoadingTile(last_station->xy);
+				}
+			}
+		}
+
+		for (Station *st : Station::Iterate()) {
+			for (size_t i = 0; i < NUM_CARGO; i++) {
+				GoodsEntry *ge = &st->goods[i];
+				for (auto it = ge->cargo.Packets()->begin(); it != ge->cargo.Packets()->end(); it++) {
+					for (CargoPacket *cp : it->second) {
+						cp->AfterLoadBeforeCargoMovement();
```

Comment:
AfterLoadBeforeCargoMovement is a copy of this whole loop, you're basically running the whole conversion quadratically. You just need to update `cp->cargo_movement here`.